File: pyvrp/cpp/crossover/make_giant_tour.cpp
#include "make_giant_tour.h"

void makeGiantTour(Individual &individual, ProblemData const *data)
{
    auto routes = individual.routes;

    // a precedes b only when a is not empty and b is. Combined with a stable
    // sort, this ensures we keep the original sorting as much as possible, but
    // also make sure all empty routes are at the end of routes_.
    auto comp = [](auto &a, auto &b) { return !a.empty() && b.empty(); };
    std::stable_sort(routes.begin(), routes.end(), comp);

    std::vector<int> tour;
    tour.reserve(data->nbClients);

    for (auto const &route : routes)
        for (Client c : route)
            tour.push_back(c);

    return tour;
};

File: pyvrp/crossover/make_giant_tour.py
from typing import List


def make_giant_tour(routes: List[List[int]]) -> List[int]:
    """
    Makes a giant tour of the passed-in routes.

    Parameters
    ----------
    routes
        The routes to be merged into a giant tour.

    Returns
    -------
    List[int]
        The giant tour representation of the given routes.
    """
    return [client for route in routes for client in route]
